napari/_vispy/visuals/volume.py

```diff
@@ -1,129 +1,132 @@
-
-            color1 = $sample( u_volumetex, loc+vec3(0.0,-step[1],0.0) );
-            color2 = $sample( u_volumetex, loc+vec3(0.0,step[1],0.0) );
-            val1 = colorToVal(color1);
-            val2 = colorToVal(color2);
-            N[1] = val1 - val2;
-            n_bg_borders += detectAdjacentBackground(val1, val2);
-
-            color1 = $sample( u_volumetex, loc+vec3(0.0,0.0,-step[2]) );
-            color2 = $sample( u_volumetex, loc+vec3(0.0,0.0,step[2]) );
-            val1 = colorToVal(color1);
-            val2 = colorToVal(color2);
-            N[2] = val1 - val2;
-            n_bg_borders += detectAdjacentBackground(val1, val2);
-
-            // Normalize and flip normal so it points towards viewer
-            N = normalize(N);
-            float Nselect = float(dot(N,V) > 0.0);
-            N = (2.0*Nselect - 1.0) * N;  // ==  Nselect * N - (1.0-Nselect)*N;
-
-            // Init colors
-            vec4 ambient_color = vec4(0.0, 0.0, 0.0, 0.0);
-            vec4 diffuse_color = vec4(0.0, 0.0, 0.0, 0.0);
-            vec4 final_color;
-
-            // todo: allow multiple light, define lights on viewvox or subscene
-            int nlights = 1;
-            for (int i=0; i<nlights; i++)
-            {
-                // Get light direction (make sure to prevent zero devision)
-                vec3 L = normalize(view_ray);  //lightDirs[i];
-                float lightEnabled = float( length(L) > 0.0 );
-                L = normalize(L+(1.0-lightEnabled));
-
-                // Calculate lighting properties
-                float lambertTerm = clamp( dot(N,L), 0.0, 1.0 );
-                if (n_bg_borders > 0) {
-                    // to fix dim pixels due to poor normal estimation,
-                    // we give a default lambda to pixels surrounded by background
-                    lambertTerm = 0.5;
-                }
-
-                // Calculate mask
-                float mask1 = lightEnabled;
-
-                // Calculate colors
-                ambient_color +=  mask1 * u_ambient;  // * gl_LightSource[i].ambient;
-                diffuse_color +=  mask1 * lambertTerm;
-            }
-
-            // Calculate final color by componing different components
-            final_color = betterColor * ( ambient_color + diffuse_color);
-            final_color.a = betterColor.a;
-
-            // Done
-            return final_color;
-        }
         // check if value is different from the background value
-        if ( floatNotEqual(val, u_categorical_bg_value, u_equality_tolerance) ) {
+        if ( floatNotEqual(val, categorical_bg_value) ) {
             // Take the last interval in smaller steps
             vec3 iloc = loc - step;
             for (int i=0; i<10; i++) {
                 color = $sample(u_volumetex, iloc);
-                if (floatNotEqual(color.g, u_categorical_bg_value, u_equality_tolerance) ) {
+                if (floatNotEqual(color.g, categorical_bg_value) ) {
```

Comment:
The sampling is broken, so color is empty. Which also means this if statement does not pass, ever. This is easy to see by adding the following line just above or below the inner loop:

```glsl
surface_found = true;
color = vec4(1);
```

This will display the labels, if with wrong colors.

Reply:
@brisvag , is the sampling still broken or was that fixed when you changed the visual back to CPU scaling?

Reply:
yes, this works now!